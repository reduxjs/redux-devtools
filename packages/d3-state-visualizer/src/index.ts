export { tree } from './charts';
export type { InputOptions, NodeWithId, Primitive } from './charts';
