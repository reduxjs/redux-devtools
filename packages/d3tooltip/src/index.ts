import d3Package, { Primitive, Selection } from 'd3';
import { is } from 'ramda';
import utils from './utils';
const { prependClass, functor } = utils;

interface Options<Datum> {
  left: number | undefined;
  top: number | undefined;
  offset: {
    left: number;
    top: number;
  };
  root: Selection<Datum> | undefined;
}

const defaultOptions: Options<unknown> = {
  left: undefined, // mouseX
  top: undefined, // mouseY
  offset: { left: 0, top: 0 },
  root: undefined,
};

interface Tip<Datum> {
  (selection: Selection<Datum>): void;
  attr: (
    this: this,
    d:
      | string
      | {
          [key: string]:
            | Primitive
            | ((datum: Datum, index: number, outerIndex: number) => Primitive);
        }
  ) => this;
  style: (
    this: this,
    d:
      | string
      | {
          [key: string]:
            | Primitive
            | ((datum: Datum, index: number, outerIndex: number) => Primitive);
        }
      | undefined
  ) => this;
  text: (
    this: this,
    d: string | ((datum: Datum, index?: number, outerIndex?: number) => string)
  ) => this;
}

export function tooltip<Datum>(
  d3: typeof d3Package,
  className = 'tooltip',
  options: Partial<Options<Datum>> = {}
): Tip<Datum> {
  const { left, top, offset, root } = {
    ...defaultOptions,
    ...options,
  } as Options<Datum>;

  let attrs = { class: className };
  let text: (datum: Datum, index?: number, outerIndex?: number) => string = (
    node: Datum
  ) => '';
  let styles = {};

  let el: Selection<Datum>;
  const anchor = root || d3.select('body');
  const rootNode = anchor.node();

  function tip(selection: Selection<Datum>) {
    selection.on('mouseover.tip', (node) => {
      const [mouseX, mouseY] = d3.mouse(rootNode);
      const [x, y] = [left || mouseX + offset.left, top || mouseY - offset.top];

      anchor.selectAll(`div.${className}`).remove();

      el = anchor
        .append('div')
        .attr(prependClass(className)(attrs))
        .style({
          position: 'absolute',
          'z-index': 1001,
          left: `${x}px`,
          top: `${y}px`,
          ...styles,
        })
        .html(() => text(node)) as Selection<Datum>;
    });

    selection.on('mousemove.tip', (node) => {
      const [mouseX, mouseY] = d3.mouse(rootNode);
      const [x, y] = [left || mouseX + offset.left, top || mouseY - offset.top];

      el.style({
        left: `${x}px`,
        top: `${y}px`,
      }).html(() => text(node));
    });

    selection.on('mouseout.tip', () => el.remove());
  }

  tip.attr = function setAttr(
    this: typeof tip,
    d:
      | string
      | {
          [key: string]:
            | Primitive
            | ((datum: Datum, index: number, outerIndex: number) => Primitive);
        }
  ) {
    if (is(Object, d)) {
      attrs = {
        ...attrs,
        ...(d as {
          [key: string]:
            | Primitive
            | ((datum: Datum, index: number, outerIndex: number) => Primitive);
        }),
      };
    }
    return this;
  };

  tip.style = function setStyle(
    this: typeof tip,
    d:
      | string
      | {
          [key: string]:
            | Primitive
            | ((datum: Datum, index: number, outerIndex: number) => Primitive);
        }
      | undefined
  ) {
    if (is(Object, d)) {
      styles = {
        ...styles,
        ...(d as {
          [key: string]:
            | Primitive
            | ((datum: Datum, index: number, outerIndex: number) => Primitive);
        }),
      };
    }
    return this;
  };

  tip.text = function setText(
    this: typeof tip,
    d: string | ((datum: Datum, index?: number, outerIndex?: number) => string)
  ) {
    text = functor(d);
    return this;
  };

  return tip;
}
