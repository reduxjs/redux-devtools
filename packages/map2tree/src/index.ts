import isArray from 'lodash/isArray';
import isPlainObject from 'lodash/isPlainObject';
import mapValues from 'lodash/mapValues';

export interface Node {
  name: string;
  children?: Node[] | null;
  value?: unknown;
}

function visit(
  parent: Node,
  visitFn: (parent: Node) => void,
  childrenFn: (parent: Node) => Node[] | undefined | null
) {
  if (!parent) return;

  visitFn(parent);

  const children = childrenFn(parent);
  if (children) {
    const count = children.length;
    for (let i = 0; i < count; i++) {
      visit(children[i], visitFn, childrenFn);
    }
  }
}

function getNode(tree: Node, key: string): Node | null {
  let node = null;

  visit(
    tree,
    (d) => {
      if (d.name === key) {
        node = d;
      }
    },
    (d) => d.children
  );

  return node;
}

export function map2tree(
  // eslint-disable-next-line @typescript-eslint/ban-types
  root: unknown,
  options: { key?: string; pushMethod?: 'push' | 'unshift' } = {},
  tree: Node = { name: options.key || 'state', children: [] }
  // eslint-disable-next-line @typescript-eslint/ban-types
): Node | {} {
  // eslint-disable-next-line @typescript-eslint/ban-types
  if (!isPlainObject(root) && root && !(root as { toJS: () => {} }).toJS) {
    return {};
  }

  const { key: rootNodeKey = 'state', pushMethod = 'push' } = options;
  const currentNode = getNode(tree, rootNodeKey);

  if (currentNode === null) {
    return {};
  }

  mapValues(
    // eslint-disable-next-line @typescript-eslint/ban-types
    root && (root as { toJS: () => {} }).toJS
      ? // eslint-disable-next-line @typescript-eslint/ban-types
        (root as { toJS: () => {} }).toJS()
      : // eslint-disable-next-line @typescript-eslint/ban-types
        (root as {}),
    // eslint-disable-next-line @typescript-eslint/ban-types
    (maybeImmutable: { toJS?: () => {} }, key) => {
      const value =
        maybeImmutable && maybeImmutable.toJS
          ? maybeImmutable.toJS()
          : maybeImmutable;
      const newNode: Node = { name: key };

      if (isArray(value)) {
        newNode.children = [];

        for (let i = 0; i < value.length; i++) {
          newNode.children[pushMethod]({
            name: `${key}[${i}]`,
            [isPlainObject(value[i]) ? 'object' : 'value']: value[i],
          });
        }
      } else if (isPlainObject(value)) {
        newNode.children = [];
      } else {
        newNode.value = value;
      }

      currentNode.children![pushMethod](newNode);

      map2tree(value, { key, pushMethod }, tree);
    }
  );

  return tree;
}
