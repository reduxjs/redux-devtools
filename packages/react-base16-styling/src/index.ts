import * as base16 from 'base16';
import { Base16Theme } from 'base16';
import Color from 'color';
import * as CSS from 'csstype';
import curry from 'lodash.curry';
import { Color as ColorTuple, yuv2rgb, rgb2yuv } from './colorConverters';
import {
  Styling,
  StylingConfig,
  StylingFunction,
  StylingValue,
  StylingValueFunction,
  Theme,
} from './types';

const DEFAULT_BASE16 = base16.default;

const BASE16_KEYS = Object.keys(DEFAULT_BASE16);

// we need a correcting factor, so that a dark, but not black background color
// converts to bright enough inversed color
const flip = (x: number) => (x < 0.25 ? 1 : x < 0.5 ? 0.9 - x : 1.1 - x);

const invertColor = (hexString: string) => {
  const color = Color(hexString);
  const [y, u, v] = rgb2yuv(color.array() as ColorTuple);
  const flippedYuv: ColorTuple = [flip(y), u, v];
  const rgb = yuv2rgb(flippedYuv);
  return Color.rgb(rgb).hex();
};

const merger = (styling: Partial<Styling>) => {
  return (prevStyling: Partial<Styling>) => ({
    className: [prevStyling.className, styling.className]
      .filter(Boolean)
      .join(' '),
    style: { ...(prevStyling.style || {}), ...(styling.style || {}) },
  });
};

const mergeStyling = (
  customStyling: StylingValue,
  defaultStyling: StylingValue
): StylingValue | undefined => {
  if (customStyling === undefined) {
    return defaultStyling;
  }
  if (defaultStyling === undefined) {
    return customStyling;
  }

  const customType = typeof customStyling;
  const defaultType = typeof defaultStyling;

  switch (customType) {
    case 'string':
      switch (defaultType) {
        case 'string':
          return [defaultStyling, customStyling].filter(Boolean).join(' ');
        case 'object':
          return merger({
            className: customStyling as string,
            style: defaultStyling as CSS.Properties<string | number>,
          });
        case 'function':
          return (styling: Styling, ...args: unknown[]) =>
            merger({
              className: customStyling as string,
            })((defaultStyling as StylingValueFunction)(styling, ...args));
      }
      break;
    case 'object':
      switch (defaultType) {
        case 'string':
          return merger({
            className: defaultStyling as string,
            style: customStyling as CSS.Properties<string | number>,
          });
        case 'object':
          return {
            ...(defaultStyling as CSS.Properties<string | number>),
            ...(customStyling as CSS.Properties<string | number>),
          };
        case 'function':
          return (styling: Styling, ...args: unknown[]) =>
            merger({
              style: customStyling as CSS.Properties<string | number>,
            })((defaultStyling as StylingValueFunction)(styling, ...args));
      }
      break;
    case 'function':
      switch (defaultType) {
        case 'string':
          return (styling, ...args) =>
            (customStyling as StylingValueFunction)(
              merger(styling)({
                className: defaultStyling as string,
              }),
              ...args
            );
        case 'object':
          return (styling, ...args) =>
            (customStyling as StylingValueFunction)(
              merger(styling)({
                style: defaultStyling as CSS.Properties<string | number>,
              }),
              ...args
            );
        case 'function':
          return (styling, ...args) =>
            (customStyling as StylingValueFunction)(
              (defaultStyling as StylingValueFunction)(
                styling,
                ...args
              ) as Styling,
              ...args
            );
      }
  }
};

const mergeStylings = (
  customStylings: StylingConfig,
  defaultStylings: StylingConfig
): StylingConfig => {
  const keys = Object.keys(defaultStylings);
  for (const key in customStylings) {
    if (keys.indexOf(key) === -1) keys.push(key);
  }

  return keys.reduce(
    (mergedStyling, key) => (
      (mergedStyling[key as keyof StylingConfig] = mergeStyling(
        customStylings[key] as StylingValue,
        defaultStylings[key] as StylingValue
      ) as StylingValue),
      mergedStyling
    ),
    {} as StylingConfig
  );
};

const getStylingByKeys = (
  mergedStyling: StylingConfig,
  keys: (string | false | undefined) | (string | false | undefined)[],
  ...args: unknown[]
): Styling => {
  if (keys === null) {
    return mergedStyling as unknown as Styling;
  }

  if (!Array.isArray(keys)) {
    keys = [keys];
  }

  const styles = keys
    .map((key) => mergedStyling[key as string])
    .filter(Boolean);

  const props = styles.reduce<Styling>(
    (obj, s) => {
      if (typeof s === 'string') {
        obj.className = [obj.className, s].filter(Boolean).join(' ');
      } else if (typeof s === 'object') {
        obj.style = { ...obj.style, ...s };
      } else if (typeof s === 'function') {
        obj = { ...obj, ...s(obj, ...args) };
      }

      return obj;
    },
    { className: '', style: {} }
  );

  if (!props.className) {
    delete props.className;
  }

  if (Object.keys(props.style!).length === 0) {
    delete props.style;
  }

  return props;
};

export const invertBase16Theme = (base16Theme: Base16Theme): Base16Theme =>
  Object.keys(base16Theme).reduce(
    (t, key) => (
      (t[key as keyof Base16Theme] = /^base/.test(key)
        ? invertColor(base16Theme[key as keyof Base16Theme])
        : key === 'scheme'
        ? base16Theme[key] + ':inverted'
        : base16Theme[key as keyof Base16Theme]),
      t
    ),
    {} as Base16Theme
  );

interface Options {
  defaultBase16?: Base16Theme;
  base16Themes?: { [themeName: string]: Base16Theme };
}

export const createStyling = curry<
  (base16Theme: Base16Theme) => StylingConfig,
  Options | undefined,
  Theme | undefined,
  StylingFunction
>(
  (
    getStylingFromBase16: (base16Theme: Base16Theme) => StylingConfig,
    options: Options = {},
    themeOrStyling: Theme = {},
    ...args
  ): StylingFunction => {
    const { defaultBase16 = DEFAULT_BASE16, base16Themes = null } = options;

    const base16Theme = getBase16Theme(themeOrStyling, base16Themes);
    if (base16Theme) {
      themeOrStyling = {
        ...base16Theme,
        ...(themeOrStyling as Base16Theme | StylingConfig),
      };
    }

    const theme = BASE16_KEYS.reduce(
      (t, key) => (
        (t[key as keyof Base16Theme] =
          (themeOrStyling as Base16Theme)[key as keyof Base16Theme] ||
          defaultBase16[key as keyof Base16Theme]),
        t
      ),
      {} as Base16Theme
    );

    const customStyling = Object.keys(themeOrStyling).reduce(
      (s, key) =>
        BASE16_KEYS.indexOf(key) === -1
          ? ((s[key] = (themeOrStyling as StylingConfig)[key]), s)
          : s,
      {} as StylingConfig
    );

    const defaultStyling = getStylingFromBase16(theme);

    const mergedStyling = mergeStylings(customStyling, defaultStyling);

    return curry(getStylingByKeys, 2)(mergedStyling, ...args);
  },
  3
);

const isStylingConfig = (theme: Theme): theme is StylingConfig =>
  !!(theme as StylingConfig).extend;

export const getBase16Theme = (
  theme: Theme,
  base16Themes?: { [themeName: string]: Base16Theme } | null
): Base16Theme | undefined => {
  if (theme && isStylingConfig(theme) && theme.extend) {
    theme = theme.extend as string | Base16Theme;
  }

  if (typeof theme === 'string') {
    const [themeName, modifier] = theme.split(':');
    if (base16Themes) {
      theme = base16Themes[themeName];
    } else {
      theme = base16[themeName as keyof typeof base16];
    }
    if (modifier === 'inverted') {
      theme = invertBase16Theme(theme);
    }
  }

  return theme && Object.prototype.hasOwnProperty.call(theme, 'base00')
    ? (theme as Base16Theme)
    : undefined;
};

export const invertTheme = (theme: Theme | undefined): Theme | undefined => {
  if (typeof theme === 'string') {
    return `${theme}:inverted`;
  }

  if (theme && isStylingConfig(theme) && theme.extend) {
    if (typeof theme.extend === 'string') {
      return { ...theme, extend: `${theme.extend}:inverted` };
    }

    return {
      ...theme,
      extend: invertBase16Theme(theme.extend as Base16Theme),
    };
  }

  if (theme) {
    return invertBase16Theme(theme as Base16Theme);
  }

  return theme;
};

export type { Base16Theme };
export * from './types';
