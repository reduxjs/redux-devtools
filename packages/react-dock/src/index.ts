export { default as Dock } from './Dock';
