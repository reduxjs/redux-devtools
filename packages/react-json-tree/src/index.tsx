// ES6 + inline style port of JSONViewer https://bitbucket.org/davevedder/react-json-viewer/
// all credits and original code to the author
// Dave Vedder <veddermatic@gmail.com> http://www.eskimospy.com/
// port by Daniele Zannotti http://www.github.com/dzannotti <dzannotti@me.com>

import React from 'react';
import PropTypes from 'prop-types';
import JSONNode from './JSONNode';
import createStylingFromTheme from './createStylingFromTheme';
import {
  invertTheme,
  type StylingConfig,
  type StylingFunction,
  type StylingValue,
  type Theme,
} from 'react-base16-styling';
import { CircularPropsPassedThroughJSONTree } from './types';

interface Props extends CircularPropsPassedThroughJSONTree {
  data: any;
  theme?: Theme;
  invertTheme: boolean;
}

interface State {
  styling: StylingFunction;
}

const identity = (value: any) => value;
const expandRootNode = (
  keyPath: (string | number)[],
  data: any,
  level: number
) => level === 0;
const defaultItemString = (
  type: string,
  data: any,
  itemType: React.ReactNode,
  itemString: string
) => (
  <span>
    {itemType} {itemString}
  </span>
);
const defaultLabelRenderer = ([label]: (string | number)[]) => (
  <span>{label}:</span>
);
const noCustomNode = () => false;

function checkLegacyTheming(theme: Theme | undefined, props: Props) {
  const deprecatedStylingMethodsMap = {
    getArrowStyle: 'arrow',
    getListStyle: 'nestedNodeChildren',
    getItemStringStyle: 'nestedNodeItemString',
    getLabelStyle: 'label',
    getValueStyle: 'valueText',
  };

  const deprecatedStylingMethods = Object.keys(
    deprecatedStylingMethodsMap
  ).filter((name) => props[name as keyof Props]);

  if (deprecatedStylingMethods.length > 0) {
    if (typeof theme === 'string') {
      theme = {
        extend: theme,
      };
    } else {
      theme = { ...theme };
    }

    deprecatedStylingMethods.forEach((name) => {
      // eslint-disable-next-line no-console
      console.error(
        `Styling method "${name}" is deprecated, use "theme" property instead`
      );

      (theme as StylingConfig)[
        deprecatedStylingMethodsMap[
          name as keyof typeof deprecatedStylingMethodsMap
        ]
      ] = ({ style }, ...args) => ({
        style: {
          ...style,
          ...props[name as keyof Props](...args),
        },
      });
    });
  }

  return theme;
}

function getStateFromProps(props: Props) {
  let theme = checkLegacyTheming(props.theme, props);
  if (props.invertTheme) {
    theme = invertTheme(theme);
  }

  return {
    styling: createStylingFromTheme(theme),
  };
}

export class JSONTree extends React.Component<Props, State> {
  static propTypes = {
    data: PropTypes.any,
    hideRoot: PropTypes.bool,
    theme: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
    invertTheme: PropTypes.bool,
    keyPath: PropTypes.arrayOf(
      PropTypes.oneOfType([PropTypes.string, PropTypes.number])
    ),
    postprocessValue: PropTypes.func,
    sortObjectKeys: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
  };

  static defaultProps = {
    shouldExpandNode: expandRootNode,
    hideRoot: false,
    keyPath: ['root'],
    getItemString: defaultItemString,
    labelRenderer: defaultLabelRenderer,
    valueRenderer: identity,
    postprocessValue: identity,
    isCustomNode: noCustomNode,
    collectionLimit: 50,
    invertTheme: true,
  };

  constructor(props: Props) {
    super(props);
    this.state = getStateFromProps(props);
  }

  UNSAFE_componentWillReceiveProps(nextProps: Props) {
    if (
      ['theme', 'invertTheme'].find(
        (k) => nextProps[k as keyof Props] !== this.props[k as keyof Props]
      )
    ) {
      this.setState(getStateFromProps(nextProps));
    }
  }

  shouldComponentUpdate(nextProps: Props) {
    return !!Object.keys(nextProps).find((k) =>
      k === 'keyPath'
        ? nextProps[k].join('/') !== this.props[k].join('/')
        : nextProps[k as keyof Props] !== this.props[k as keyof Props]
    );
  }

  render() {
    const {
      data: value,
      keyPath,
      postprocessValue,
      hideRoot,
      theme, // eslint-disable-line no-unused-vars
      invertTheme: _, // eslint-disable-line no-unused-vars
      ...rest
    } = this.props;

    const { styling } = this.state;

    return (
      <ul {...styling('tree')}>
        <JSONNode
          {...{ postprocessValue, hideRoot, styling, ...rest }}
          keyPath={hideRoot ? [] : keyPath}
          value={postprocessValue(value)}
        />
      </ul>
    );
  }
}

export { StylingValue };
