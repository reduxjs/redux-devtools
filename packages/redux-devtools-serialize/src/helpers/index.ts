export interface SerializedData {
  data: unknown;
  __serializedType__: string;
  __serializedRef__?: number;
}

export function mark(data: unknown, type: string): SerializedData;
export function mark<K extends string>(
  data: { [key in K]: () => unknown },
  type: string,
  transformMethod?: K | false
): SerializedData;
export function mark<K extends string>(
  data: any,
  type: string,
  transformMethod?: 'toString' | false
): SerializedData;
export function mark<K extends string>(
  data: { [key in K]: () => unknown } | unknown,
  type: string,
  transformMethod?: K | false
): SerializedData {
  return {
    data: transformMethod
      ? (data as { [key in K]: () => unknown })[transformMethod]()
      : data,
    __serializedType__: type,
  };
}

export function extract(data: unknown, type: string): SerializedData {
  return {
    data: Object.assign({}, data),
    __serializedType__: type,
  };
}

export function refer(data: unknown, type: string): SerializedData;
export function refer<K extends string>(
  data: { [key in K]: () => unknown },
  type: string,
  transformMethod?: K | false,
  refs?: (new (data: any) => unknown)[] | null
): SerializedData;
export function refer<K extends string>(
  data: any,
  type: string,
  transformMethod?: 'toString' | false,
  refs?: (new (data: any) => unknown)[] | null
): SerializedData;
export function refer<K extends string>(
  data: { [key in K]: () => unknown } | unknown,
  type: string,
  transformMethod?: K | false,
  refs?: (new (data: any) => unknown)[] | null
): SerializedData {
  const r = mark(data as { [key in K]: () => unknown }, type, transformMethod);
  if (!refs) return r;
  for (let i = 0; i < refs.length; i++) {
    const ref = refs[i];
    if (typeof ref === 'function' && data instanceof ref) {
      r.__serializedRef__ = i;
      return r;
    }
  }
  return r;
}
