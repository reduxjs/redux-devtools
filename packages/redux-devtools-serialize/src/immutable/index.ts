import jsan from 'jsan';
import Immutable from 'immutable';
import serialize from './serialize';
import options from '../constants/options';

export default function (
  immutable: typeof Immutable,
  refs?: (new (data: any) => unknown)[] | null,
  customReplacer?: (
    key: string,
    value: unknown,
    defaultReplacer: (key: string, value: unknown) => unknown,
  ) => unknown,
  customReviver?: (
    key: string,
    value: unknown,
    defaultReviver: (key: string, value: unknown) => unknown,
  ) => unknown,
) {
  return {
    stringify: function (data: unknown) {
      return jsan.stringify(
        data,
        serialize(immutable, refs, customReplacer, customReviver).replacer,
        undefined,
        options,
      );
    },
    parse: function (data: string) {
      return jsan.parse(
        data,
        serialize(immutable, refs, customReplacer, customReviver).reviver,
      );
    },
    serialize: serialize,
  };
}
export { default as serialize } from './serialize';
