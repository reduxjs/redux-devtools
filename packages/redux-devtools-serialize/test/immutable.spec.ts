import Immutable, { Map, OrderedMap } from 'immutable';
import Serialize from '../src/immutable';
import { SerializedData } from '../src/helpers';
const serialize = Serialize(Immutable);
const stringify = serialize.stringify;
const parse = serialize.parse;

const data = {
  map: Immutable.Map({ a: 1, b: 2, c: 3, d: 4 }),
  orderedMap: Immutable.OrderedMap({ b: 2, a: 1, c: 3, d: 4 }),
  list: Immutable.List([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]),
  range: Immutable.Range(0, 7),
  repeat: Immutable.Repeat('hi', 100),
  set: Immutable.Set([10, 9, 8, 7, 6, 5, 4, 3, 2, 1]),
  orderedSet: Immutable.OrderedSet([10, 9, 8, 7, 6, 5, 4, 3, 2, 1]),
  seq: Immutable.Seq([1, 2, 3, 4, 5, 6, 7, 8]),
  stack: Immutable.Stack.of('a', 'b', 'c'),
};

describe('Immutable', function () {
  const stringified: { [key: string]: string } = {};
  describe('Stringify', function () {
    Object.keys(data).forEach(function (key) {
      // eslint-disable-next-line jest/valid-title
      it(key, function () {
        stringified[key] = stringify(data[key as keyof typeof data]);
        expect(stringified[key]).toMatchSnapshot();
      });
    });
  });

  describe('Parse', function () {
    Object.keys(data).forEach(function (key) {
      // eslint-disable-next-line jest/valid-title
      it(key, function () {
        expect(parse(stringified[key])).toEqual(data[key as keyof typeof data]);
      });
    });
  });

  describe('Record', function () {
    const ABRecord = Immutable.Record({ a: 1, b: 2 });
    const myRecord = new ABRecord({ b: 3 });

    const serialize = Serialize(Immutable, [ABRecord]);
    const stringify = serialize.stringify;
    const parse = serialize.parse;
    let stringifiedRecord: string;

    it('stringify', function () {
      stringifiedRecord = stringify(myRecord);
      expect(stringifiedRecord).toMatchSnapshot();
    });

    it('parse', function () {
      expect(parse(stringifiedRecord)).toEqual(myRecord);
    });
  });

  describe('Nested', function () {
    const ABRecord = Immutable.Record({
      map: Immutable.OrderedMap({ seq: data.seq, stack: data.stack }),
      repeat: data.repeat,
    });
    const nestedData = Immutable.Set([ABRecord(), data.orderedSet, data.range]);

    const serialize = Serialize(Immutable, [ABRecord]);
    const stringify = serialize.stringify;
    const parse = serialize.parse;
    let stringifiedNested: string;

    it('stringify', function () {
      stringifiedNested = stringify(nestedData);
      expect(stringifiedNested).toMatchSnapshot();
    });

    it('parse', function () {
      expect(parse(stringifiedNested)).toEqual(nestedData);
    });
  });
  describe('With references', function () {
    it('serializes and deserializes', function () {
      const sharedValue: unknown[] = [];
      const record = Immutable.Record({
        prop: sharedValue,
      });

      const refs = [record];

      const obj = Immutable.Map({
        fst: new record(),
        scnd: new record(),
      });

      const serialize = Serialize(Immutable, refs);
      const serialized = serialize.stringify(obj);
      const parsed = JSON.parse(serialized);

      const fstProp = parsed.data.fst.data.prop;
      const scndProp = parsed.data.scnd.data.prop;

      expect(fstProp).toEqual(scndProp);
      expect(Array.isArray(obj.get('fst')!.get('prop'))).toBe(true);
    });
  });

  describe('Custom replacer and reviver functions', function () {
    const customOneRepresentation = 'one';

    function customReplacer(
      key: string,
      value: unknown,
      defaultReplacer: (key: string, value: unknown) => unknown,
    ) {
      if (value === 1) {
        return { data: customOneRepresentation, __serializedType__: 'number' };
      }
      return defaultReplacer(key, value);
    }

    function customReviver(
      key: string,
      value: unknown,
      defaultReviver: (key: string, value: unknown) => unknown,
    ) {
      if (
        typeof value === 'object' &&
        (value as SerializedData).__serializedType__ === 'number' &&
        (value as SerializedData).data === customOneRepresentation
      ) {
        return 1;
      }
      return defaultReviver(key, value);
    }

    const serializeCustom = Serialize(
      Immutable,
      null,
      customReplacer,
      customReviver,
    );

    Object.keys(data).forEach(function (key) {
      const stringified = serializeCustom.stringify(
        data[key as keyof typeof data],
      );
      // eslint-disable-next-line jest/valid-title
      it(key, function () {
        const deserialized = serializeCustom.parse(stringified);
        expect(deserialized).toEqual(data[key as keyof typeof data]);
        if (key === 'map' || key === 'orderedMap') {
          const deserializedDefault = parse(stringified);
          // eslint-disable-next-line jest/no-conditional-expect
          expect(
            (
              deserializedDefault as
                | Map<unknown, unknown>
                | OrderedMap<unknown, unknown>
            ).get('a'),
          ).toEqual(customOneRepresentation);
        }
      });
    });
  });
});
