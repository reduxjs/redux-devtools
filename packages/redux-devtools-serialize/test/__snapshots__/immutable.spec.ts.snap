// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Immutable Nested stringify 1`] = `"{"data":[{"data":{"map":{"data":{"seq":{"data":[1,2,3,4,5,6,7,8],"__serializedType__":"ImmutableSeq"},"stack":{"data":["a","b","c"],"__serializedType__":"ImmutableStack"}},"__serializedType__":"ImmutableOrderedMap"},"repeat":{"data":{"_value":"hi","size":100},"__serializedType__":"ImmutableRepeat"}},"__serializedType__":"ImmutableRecord","__serializedRef__":0},{"data":[10,9,8,7,6,5,4,3,2,1],"__serializedType__":"ImmutableOrderedSet"},{"data":{"_start":0,"_end":7,"_step":1,"size":7},"__serializedType__":"ImmutableRange"}],"__serializedType__":"ImmutableSet"}"`;

exports[`Immutable Record stringify 1`] = `"{"data":{"a":1,"b":3},"__serializedType__":"ImmutableRecord","__serializedRef__":0}"`;

exports[`Immutable Stringify list 1`] = `"{"data":[1,2,3,4,5,6,7,8,9,10],"__serializedType__":"ImmutableList"}"`;

exports[`Immutable Stringify map 1`] = `"{"data":{"a":1,"b":2,"c":3,"d":4},"__serializedType__":"ImmutableMap"}"`;

exports[`Immutable Stringify orderedMap 1`] = `"{"data":{"b":2,"a":1,"c":3,"d":4},"__serializedType__":"ImmutableOrderedMap"}"`;

exports[`Immutable Stringify orderedSet 1`] = `"{"data":[10,9,8,7,6,5,4,3,2,1],"__serializedType__":"ImmutableOrderedSet"}"`;

exports[`Immutable Stringify range 1`] = `"{"data":{"_start":0,"_end":7,"_step":1,"size":7},"__serializedType__":"ImmutableRange"}"`;

exports[`Immutable Stringify repeat 1`] = `"{"data":{"_value":"hi","size":100},"__serializedType__":"ImmutableRepeat"}"`;

exports[`Immutable Stringify seq 1`] = `"{"data":[1,2,3,4,5,6,7,8],"__serializedType__":"ImmutableSeq"}"`;

exports[`Immutable Stringify set 1`] = `"{"data":[1,2,3,4,5,6,7,8,9,10],"__serializedType__":"ImmutableSet"}"`;

exports[`Immutable Stringify stack 1`] = `"{"data":["a","b","c"],"__serializedType__":"ImmutableStack"}"`;
